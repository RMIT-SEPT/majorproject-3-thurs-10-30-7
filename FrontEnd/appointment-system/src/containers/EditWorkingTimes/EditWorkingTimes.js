import React, { Component } from "react";
import { connect } from "react-redux";
import axios from "../../axios-sept";
import moment from "moment";
import classes from "./EditWorkingTimes.module.css";
import { withRouter } from "react-router-dom";

class EditWorkingTimes extends Component {
  state = {
    loading: false,
    error: null,
    startTime: moment((new URLSearchParams(window.location.search)).get('startTime')).format("H"),
    endTime: moment((new URLSearchParams(window.location.search)).get('endTime')).format("H"),
    defaultStartTime: (new URLSearchParams(window.location.search)).get('startTime'),
    defaultEndTime: (new URLSearchParams(window.location.search)).get('endTime'),
    employeeId: (new URLSearchParams(window.location.search)).get('employeeId'),
    workingTimeId: (new URLSearchParams(window.location.search)).get('workingTimeId'),
  }

 

  render() {

    const updateWorkingTimes = async () => {
      this.setState({...this.state, loading: true, error: null });
      const config = {
        headers: {
          Authorization: `Bearer ${this.props.token}`,
        },
      };
  
      try {
        if (
          parseInt(this.state.startTime, 10) > parseInt(this.state.endTime, 10)
        ) {
          this.setState({ error: `End Time has to be before start time`, loading: false });
        } else {
          let startTimeText = this.state.startTime;
          let endTimeText = this.state.endTime;
          if (startTimeText < 10) {
            startTimeText = `0${this.state.startTime}`;
          }
          if (endTimeText < 10) {
            endTimeText = `0${this.state.endTime}`;
          }
          const baseTime = moment(this.state.defaultStartTime).format('YYYY-MM-DD');
          const data = {
            workingTimeId: this.state.workingTimeId,
            startTime: `${baseTime}T${startTimeText}:00:00`,
            endTime: `${baseTime}T${endTimeText}:00:00`,
          };
          await axios.put(`/api/workingTime`, data, config)
          .then(()=>{this.props.history.push("/viewworkingtimes")});
        }
      } catch (error) {
        this.setState({
          error: "Error updating workingTimes. Either this updated time conflicts with the employee's availabilities, or one of their bookings.",
          loading: false,
        });
      }
    };
    

    const shiftTimes = () => {
      const times = [];
      for (let i = 0; i <= 24; i++) {
        times.push(i);
      }
      return times.map((time) => (
        <option value={time} key={time}>
          {time}:00
        </option>
      ));
    };
    const setTime = (e) => {
      e.preventDefault();
      this.setState({
        [e.target.id]: e.target.value,
      });
    };

    return (
      <div className={"container " + classes.Display}>
        <h1 className='text-center'>Edit Working Times</h1>
        <div className="d-flex">
          <div className="container">
            <div className="row">
              <div className="col">Date:</div>
              <div className="col">
                {moment(this.state.defaultStartTime).format("DD/MM/yyyy")}
              </div>
            </div>
            <div className="row">
              <div className="col">Start Time:</div>
              <div className="col">
                <select
                  id='startTime'
                  className='custom-select'
                  onChange={setTime}
                  value={this.state.startTime || undefined}
                >
                  {shiftTimes()}
                </select>
              </div>
            </div>
            <div className="row">
              <div className="col">End Time:</div>
              <div className="col">
                <select
                  id='endTime'
                  className='custom-select'
                  onChange={setTime}
                  value={this.state.endTime || undefined}
                >
                  {shiftTimes()}
                </select>
              </div>
            </div>
            <div className="row">
              <button type="button" className="btn btn-primary" onClick={updateWorkingTimes}>Edit</button>
            </div>
          </div>
        </div>
        <div className="text-center"><p>{this.state.error}</p></div>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    token: state.auth.token,
    userType: state.auth.authority,
    userId: state.auth.userId,
  };
};

export default connect(mapStateToProps)(withRouter(EditWorkingTimes));
