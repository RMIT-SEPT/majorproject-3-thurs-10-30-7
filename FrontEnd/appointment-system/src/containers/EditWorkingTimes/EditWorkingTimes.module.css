.filter {
  float: left;
  width: 15%;
  margin-right: 20px;
}

.Display a[aria-controls] {
  color: #4873a6;
  font-weight: bold;
}

.Display button[type='submit'] {
  margin: 5px;
}

.Display {
  margin-top: 3%;
  background-color: white;
  padding-top: 3%;
  padding-bottom: 3%;
  padding-left: 2.5%;
  border-radius: 15px;
  background-color: white;
  margin-bottom: 3%;
}

.Display h1 {
  margin-bottom: 5%;
  font-weight: bold;
}

@media (max-width: 768px) {
  .Display {
    overflow-x: scroll;
  }
}
