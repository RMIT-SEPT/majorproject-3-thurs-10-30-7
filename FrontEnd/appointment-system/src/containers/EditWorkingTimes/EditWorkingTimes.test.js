import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { EditWorkingTimes } from './EditWorkingTimes';

configure({ adapter: new Adapter() });

describe('<EditWorkingTimes/>', () => {
  let wrapper = shallow(<EditWorkingTimes />);

  it("should render EditWorkingTimes page", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render 6 columns', () => {
    expect(wrapper.find('.col')).toHaveLength(6);
  });

  it('should render 4 rows', () => {
    expect(wrapper.find('.row')).toHaveLength(4);
  });
});
