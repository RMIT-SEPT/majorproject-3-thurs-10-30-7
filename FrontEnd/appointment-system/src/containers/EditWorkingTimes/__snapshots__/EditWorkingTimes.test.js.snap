// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditWorkingTimes/> should render EditWorkingTimes page 1`] = `ShallowWrapper {}`;
